text/0314-cdk-construct-hub.md

```diff
@@ -362,6 +367,96 @@ pipeline.
 The details of URLs to be configrued in CloudFront is available in [appendix
 "CloudFront URLs"](#cloudfront-urls).
 
+#### Backend X Frontend Compatibility
+
+As described earlier, the front-end application consumes artifacts that are stored on the backend. Since these artifacts are subject to change,
+we need to define a mechanism by which the front-end is able to detect and operate on various versions of them.
+
+To that end, we propose a generic mechanism to apply to all artifact types:
+
+Every artifact will comply to the [SemVer](https://semver.org/spec/v2.0.0.html) specification
+and include its major version in the `version` field of the schema:
+
+```json
+{
+  "version": "v1",
+  ...
+}
+```
+
+```json
+{
+  "version": "v2",
+  ...
+}
+```
+
+Every time a change in the artifact occurs, the contents of the artifact is replaced with the new version.
+This means that at any given point in time, the artifact can either be in its new version, or the old one.
+
+As a general guideline, we *prefer displaying an older version of an artifact, than to display nothing at all.*.
+To that end, the front-end application will have to support displaying all versions of the artifact.
+
+Specifically, we describe how to implement this for each type of artifact.
+
+##### Package Documentation
+
+Package documentation is created by the **Doc-Gen** function, which produces a JSON definition file:
+
+`packages/${assembly.name}/v${assembly.version}/docs-${lang}.json`
+
+Since the front-end fetches and parses this file, we'd like for it to operate on type safe interfaces that describe the schema.
+This will ensure that breaking changes will manifest as compile time errors, and force us to handle them.
+
+To that end, we create a dedicated package, called `construct-hub-docgen`, which will define this schema, and provide an API to produce it.
```

Comment:
`@construct-hub/package-description-schema`? It should be named after what it represents, not after what produces it.

Reply:
The idea was for the package to define both the schema, and expose the generator that produces that artifact based on that schema. So it is a docgenerator. 

Regarding the `@construct-hub` scope, I don't know that we need it...actually, what if we just push this as a new major version of [`jsii-docgen`](https://github.com/cdklabs/jsii-docgen)? Its essentially what it is, apart from the `constructs` specific code which we can maybe allow using a `constructsAware` option? @eladb what do you think? 99% of this code is going to be applicable outside of construct-hub.

Reply:
I'm a little skeptical about the `docgen` part of the name, because most people would likely expect this produces a docsite, not a bunch of JSON files. But the naming is not important here, as long as we agree on the responsibilities & outcomes.

Reply:
So actually I believe that even when we introduce the schema, we will still retain the option for generating a markdown file. Its already there and can actually serve as useful feature outside the scope of construct-hub. For example for generating an API reference on the GitHub repo, the same thing `jsii-docgen` does now. 

Given that, lets push it to `jsii-docgen`? I agree `docgen` might have an implicit context of `docsite`, but we can live with that, especially since we already do.

Reply:
Yes, I love it. Agree with @iliapolo that docgen should have an option to output markdown (i.e. render markdown from the doc model).

And we are slowly moving towards a future where JSII knows about constructs anyway (think c8), so this is definitely not a break in the layers. I am comfortable for jsii-docgen to know about constructs without any additional switches.